base-station/ir_interface.py

```diff
@@ -47,15 +45,377 @@
+1 byte - PacketSequence.
+1 byte - 0x01 for failure, 0x02 for success.
+
+"""
+
+import serial
+import enum
+from config import Config
+import asyncio
+import time
+from dataclasses import dataclass
+from typing import Optional
+
+# PacketType enum
+class PT(enum.Enum):
+    QTQ = b'\x01' # QueueTxBufferRequest
+    QTR = b'\x81' # QueueTxBufferResponse
+    RRQ = b'\x02' # RetrieveRxBufferRequest
+    RRR = b'\x82' # RetrieveRxBufferResponse
+    GSQ = b'\x03' # GetStatusRequest
+    GSR = b'\x83' # GetStatusResponse
+    SSQ = b'\x04' # SendStatusRequest
+    SSR = b'\x84' # SendStatusResponse
+    PRQ = b'\x05' # PopRxBufferRequest
+    PRR = b'\x85' # PopRxBufferResponse
+
+# Packet content format
+@dataclass
+class Packet:
+    seq: Optional[bytes] = None
+    packet_size: Optional[bytes] = None
+    is_success: Optional[bytes] = None
+    status: Optional[bytes] = None
+    payload: Optional[bytes] = None
+
+    def __bytes__(self):
+        data = b''
+        for b in (self.seq, self.packet_size, self.is_success, self.status, self.payload):
+            if b == None:
+                continue
+            if type(b) == int:
+                data += bytes([b])
+            else:
+                data += b
+        return data
+
+
 class IrInterface:
-    def __init__(self):
-        pass
+    preamble = b'\xD5\x55\x55\x55\x55\x55\x55\x55'
+    success = b'\x02'
+    failure = b'\x01'
+
+    def __init__(self, config: Config):
+        self.port = config.get(key="usb_port")
+        self.baudrate = config.get(key="usb_baudrate")
+        self.timeout = config.get(key="usb_timeout")
+        self.failure_try = config.get(key="usb_failure_try")
+        self.failure_wait = config.get(key="usb_failure_wait")
+        self.packet_que_max = config.get(key="packet_que_max")
+        self.duplex = config.get(key="duplex")
+        self.seq_set = set()
+        self.current_seq = 0
+        self.pt_seq_que_table = dict([(e, dict()) for e in (PT.QTR,)] + 
+            [(e, asyncio.Queue(maxsize=self.packet_que_max)) for e in (PT.RRR, PT.GSR)])
+        self.write_lock = asyncio.Lock()
+        self.read_lock = asyncio.Lock()
+        self.half_duplex_lock = asyncio.Lock()
+        self.badge_status: Packet = None
+
+        try:
+            self.serial = serial.Serial(
+                port=self.port,
+                baudrate=self.baudrate,
+                timeout=self.timeout,
+                write_timeout=self.timeout
+            )
+        except serial.SerialException as e:
+            print(f"Failed to initialize serial port: {e}")
+            raise
 
-    async def trigger_send_packet(self, data: bytes) -> bool:
+    async def trigger_send_packet(self, data: bytes, packet_type = PT.QTQ) -> bool | Packet:
         # Triggers the ir interface to send a packet.
         # Returns True if sent successfully, False otherwise.
-        pass
+        for _ in range(self.failure_try):          
+            if not self.serial.is_open:
```

Comment:
Consider making this class the same as an opened session of serial, or, make it a state of this class. Therefore, single packet operations always assumes the serial is open or it'll raise an error.